zcash_primitives/src/transaction/builder.rs

```diff
@@ -706,16 +744,167 @@ impl<P: consensus::Parameters, U: sapling::builder::ProverProgress> Builder<'_,
+        };
+
+        let unauthed_tx: TransactionData<Coinbase> = TransactionData {
+            version: TxVersion::suggested_for_branch(consensus_branch_id),
+            consensus_branch_id,
+            // We don't use lock_time in coinbase transactions.
+            lock_time: 0,
+            expiry_height: self.expiry_height,
+            #[cfg(all(
+                any(zcash_unstable = "nu7", zcash_unstable = "zfuture"),
+                feature = "zip-233"
+            ))]
+            zip233_amount: self.zip233_amount,
+            transparent_bundle: Some(transparent_bundle.clone()),
+            // # Consensus
+            //
+            // > A coinbase transaction MUST NOT have any JoinSplit descriptions.
+            //
+            // > A coinbase transaction MUST NOT have any Spend descriptions.
+            //
+            // <https://zips.z.cash/protocol/protocol.pdf#txnconsensus>
+            sprout_bundle: None,
+            sapling_bundle,
+            orchard_bundle,
+            #[cfg(zcash_unstable = "zfuture")]
+            tze_bundle: None,
+        };
+
+        let transparent_bundle =
+            transparent_bundle.map_authorization(transparent::builder::Coinbase);
+
+        let shielded_sig_commitment = signature_hash(
+            &unauthed_tx,
+            &SignableInput::Shielded,
+            &unauthed_tx.digest(TxIdDigester),
+        );
+
+        let sapling_bundle = unauthed_tx
+            .sapling_bundle
+            .map(|b| b.apply_signatures(&mut rng, *shielded_sig_commitment.as_ref(), &[]))
+            .transpose()
+            .map_err(Error::SaplingBuild)?;
+
+        let orchard_bundle = unauthed_tx
+            .orchard_bundle
+            .map(|b| {
+                b.create_proof(&orchard::circuit::ProvingKey::build(), &mut rng)
+                    .and_then(|b| {
+                        b.apply_signatures(&mut rng, *shielded_sig_commitment.as_ref(), &[])
+                    })
+            })
+            .transpose()
+            .map_err(Error::OrchardBuild)?;
+
+        let authed_tx: TransactionData<Authorized> = TransactionData {
+            version: unauthed_tx.version,
+            consensus_branch_id: unauthed_tx.consensus_branch_id,
+            lock_time: unauthed_tx.lock_time,
+            expiry_height: unauthed_tx.expiry_height,
+            #[cfg(all(
+                any(zcash_unstable = "nu7", zcash_unstable = "zfuture"),
+                feature = "zip-233"
+            ))]
+            zip233_amount: self.zip233_amount,
+            transparent_bundle: Some(transparent_bundle),
+            sprout_bundle: unauthed_tx.sprout_bundle,
+            sapling_bundle,
+            orchard_bundle,
+            #[cfg(zcash_unstable = "zfuture")]
+            tze_bundle: None,
+        };
+
+        Ok(BuildResult {
+            transaction: authed_tx.freeze().unwrap(),
+            sapling_meta,
+            orchard_meta,
+        })
     }
```

Comment:
This helper is almost identical to the (now-renamed) `Builder::build_standard`. I am not convinced that it needs to be a new method; I think the coinbase-specific logic could be handled pretty simply with closures.

I will refactor the branch in this fashion.